File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for ch in s:
            if ch in "({[":
                stack.append(ch)
            else:
                if not stack:
                    return False
                last = stack.pop()  # pop once
                if (ch == ')' and last != '(') or \
                   (ch == '}' and last != '{') or \
                   (ch == ']' and last != '['):
                    return False
        return not stack

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        int n = s.length();
        char[] stack = new char[n];  
        int top = -1; 

        for(int i=0; i<n; i++){
             char ch = s.charAt(i);

             // if opening bracket --> push
             if(ch=='(' || ch=='{' || ch=='['){
                stack[++top] = ch;   
             }

             // if closing bracket --> check top
             else{
                if(top == -1){
                    return false;  // no matching opening
                }
                // access top element of stack
                char last = stack[top--];  
                if(ch==')' && last!='(' || ch=='}' && last!='{' || ch==']' && last!='['){
                    return false;
                }
            }
        }
        return top == -1;
    }
}